### PR TITLE
Add Kingdom-internal Panel Match protos.

### PR DESCRIPTION
**Pull request by efoxepstein**
These are used by the data-layer microservices as an abstraction around
the database.

See issue #3.

---
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7)

### CODE REVIEW

File: src/main/proto/wfa/measurement/internal/kingdom/exchange.proto [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+package wfa.measurement.internal.kingdom;
+
+import "google/type/date.proto";
+import "src/main/proto/wfa/measurement/internal/kingdom/exchange_details.proto";
+
+option java_package = "org.wfanet.measurement.internal.kingdom";
+option java_multiple_files = true;
+
+// Proto representation of the internal Exchange entity type.
```
Comment (kungfucraig) [inline src/main/proto/wfa/measurement/internal/kingdom/exchange.proto:25]: Can you document what an Exchange is? 

For example:

An Exchange is a record of a particular invocation of a panelist data exchange between a panel provide and an EDP.

File: src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+package wfa.measurement.internal.kingdom;
+
+import "google/type/date.proto";
+import "src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto";
+
+option java_package = "org.wfanet.measurement.internal.kingdom";
+option java_multiple_files = true;
+
+// Proto representation of the internal ExchangeStepAttempt entity type.
```
Comment (kungfucraig) [inline src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto:25]: Same thing here. A more comprehensive description should be added.

File: src/main/proto/wfa/measurement/internal/kingdom/model_provider.proto [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+syntax = "proto3";
+
+package wfa.measurement.internal.kingdom;
+
+option java_package = "org.wfanet.measurement.internal.kingdom";
+option java_multiple_files = true;
+
+// Proto representation of the internal ModelProvider entity type.
+message ModelProvider {
```
Comment (kungfucraig) [inline src/main/proto/wfa/measurement/internal/kingdom/model_provider.proto:23]: Here too.

File: src/main/proto/wfa/measurement/internal/kingdom/recurring_exchange.proto [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+package wfa.measurement.internal.kingdom;
+
+import "src/main/proto/wfa/measurement/internal/kingdom/recurring_exchange_details.proto";
+
+option java_package = "org.wfanet.measurement.internal.kingdom";
+option java_multiple_files = true;
+
+// Proto representation of the internal RecurringExchange entity type.
+message RecurringExchange {
```
Comment (kungfucraig) [inline src/main/proto/wfa/measurement/internal/kingdom/recurring_exchange.proto:25]: And here.

File: src/main/proto/wfa/measurement/internal/kingdom/exchange_details.proto [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+syntax = "proto3";
+
+package wfa.measurement.internal.kingdom;
+
+option java_package = "org.wfanet.measurement.internal.kingdom";
+option java_multiple_files = true;
+
+// Stores metadata about an Exchange.
+message ExchangeDetails {
```
Comment (kungfucraig) [inline src/main/proto/wfa/measurement/internal/kingdom/exchange_details.proto:23]: Any particular reason you don't just put this in the same file with Exchange? Similar question for other message below.

File: src/main/proto/wfa/measurement/internal/kingdom/exchange_details.proto [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  // SHA256 hash of the audit trail, if there is one.
+  bytes audit_trail_hash = 1;
```
Comment (kungfucraig) [inline src/main/proto/wfa/measurement/internal/kingdom/exchange_details.proto:25]: This field needs more detail.

File: src/main/proto/wfa/measurement/internal/kingdom/exchange.proto [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+message Exchange {
+  fixed64 external_recurring_exchange_id = 1;
+  fixed64 external_exchange_id = 2;
+  google.type.Date date = 3;
```
Comment (kungfucraig) [inline src/main/proto/wfa/measurement/internal/kingdom/exchange.proto:29]: We don't want to keep the time as well?

File: src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+message ExchangeStepAttempt {
+  fixed64 external_recurring_exchange_id = 1;
+  fixed64 external_exchange_id = 2;
+  google.type.Date date = 3;
+  int32 step_index = 4;
```
Comment (kungfucraig) [inline src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto:30]: What does this refer to?

File: src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    SUCCEEDED = 3;
+
+    // The attempt has failed. The step should be retried. Terminal state.
+    FAILED = 4;
+  }
+
+  // State of the ExchangeStepAttempt.
+  State state = 3;
+
+  message DebugLog {
```
Comment (kungfucraig) [inline src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto:43]: Where do you intend to write this?

**Review by kungfucraig — CHANGES_REQUESTED** [review]

File: src/main/proto/wfa/measurement/internal/kingdom/exchange_details.proto (hunk 5 above)
Comment (efoxepstein) [reply src/main/proto/wfa/measurement/internal/kingdom/exchange_details.proto:23]: Our convention is to separate them. There are actually build targets that need the Details but not the entity itself.

File: src/main/proto/wfa/measurement/internal/kingdom/exchange.proto (hunk 7 above)
Comment (efoxepstein) [reply src/main/proto/wfa/measurement/internal/kingdom/exchange.proto:29]: We have no known use cases that require granularity finer than on a per-day basis.

File: src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto (hunk 9 above)
Comment (efoxepstein) [reply src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto:43]: Can you clarify your question?

**Review by SanjayVas — DISMISSED** [review]
> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7)* status: all files reviewed, 3 unresolved discussions (waiting on @efoxepstein and @kungfucraig)

___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange.proto, line 31 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7#-MXdGkbKE6pmpS3s3p-P:-MXdGkbL2PTPJCnMYFWz:bsrk2ma) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/962edf9669bcb1a58dcd9e6a4e31eac06c700372/src/main/proto/wfa/measurement/internal/kingdom/exchange.proto#L31)):*
> ```protobuf
> 
>   // The date that the `Exchange` is supposed to occur.
>   google.type.Date date = 3;
> ```

nit: skipped field 2
___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto, line 31 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7#-MXdGxRTA7Ps8hMsWGd6:-MXdGxRTA7Ps8hMsWGd7:b-sj5uym) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/962edf9669bcb1a58dcd9e6a4e31eac06c700372/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto#L31)):*
> ```protobuf
> 
>   // Date of the `Exchange`.
>   google.type.Date date = 3;
> ```

Is this just denormalized from the `Exchange`?
___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto, line 37 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7#-MXdHHMA0zwkcDq_m0Wn:-MXdHHMA0zwkcDq_m0Wo:b-uoa2zr) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/962edf9669bcb1a58dcd9e6a4e31eac06c700372/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto#L37)):*
> ```
>   // Which attempt this is. The first attempt should have index 0 and each
>   // subsequent attempt should increase the last `attempt_index` by 1.
>   int32 attempt_index = 5;
> ```
Is there an attempts list somewhere? Just trying to figure out why this is a 0-based `attempt_index` rather than a 1-based `attempt_number`.

File: src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto (hunk 9 above)
Comment (kungfucraig) [reply src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto:43]: Talked offline. Message will be written to the database.

File: src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+  // Each of these is an encrypted SignedData. Each SignedData holds a
+  // serialized SharedOutput protocol buffer from the public API.
+  repeated bytes shared_outputs = 5;
+
+  // When the ExchangeStepAttempt was created.
+  google.protobuf.Timestamp start_time = 6;
+
+  // When the ExchangeStepAttempt was last updated.
+  google.protobuf.Timestamp update_time = 7;
+}
```
Comment (kungfucraig) [inline src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto:63]: Please have newline at EoF.

**Review by kungfucraig — DISMISSED** [review]

**Review by efoxepstein — COMMENTED** [review]
> Dismissed @kungfucraig from a discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7)* status: all files reviewed, 2 unresolved discussions (waiting on @SanjayVas)

___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange.proto, line 31 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7#-MXdGkbKE6pmpS3s3p-P:-MXh5SBlESuLPGl3ehmp:bb6un1j) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/962edf9669bcb1a58dcd9e6a4e31eac06c700372/src/main/proto/wfa/measurement/internal/kingdom/exchange.proto#L31)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

nit: skipped field 2
</blockquote></details>

Done
___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto, line 31 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7#-MXdGxRTA7Ps8hMsWGd6:-MXh0q2l030IMZVTjGql:brssvce) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/962edf9669bcb1a58dcd9e6a4e31eac06c700372/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto#L31)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Is this just denormalized from the `Exchange`?
</blockquote></details>

The Exchange PK is external_recurring_exchange_id plus Date. Does that seem reasonable?
___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto, line 37 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7#-MXdHHMA0zwkcDq_m0Wn:-MXh16S9295Gbje5G_I7:b-6hf8dp) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/962edf9669bcb1a58dcd9e6a4e31eac06c700372/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto#L37)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

> ```
>   // Which attempt this is. The first attempt should have index 0 and each
>   // subsequent attempt should increase the last `attempt_index` by 1.
>   int32 attempt_index = 5;
> ```
Is there an attempts list somewhere? Just trying to figure out why this is a 0-based `attempt_index` rather than a 1-based `attempt_number`.
</blockquote></details>

No, it's arbitrary. Is starting at 1 and using attempt_number more canonical?
___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto, line 63 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7#-MXdGU72doxHXBexv3XN-r1-63:-MXh51au2l-Kr6avmyrU:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/962edf9669bcb1a58dcd9e6a4e31eac06c700372/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto#L63)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

Please have newline at EoF.
</blockquote></details>

Done.

**Review by SanjayVas — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7)* status: all files reviewed, 2 unresolved discussions (waiting on @AlekseiS, @efoxepstein, @oliver-amzn, and @zasadnyy)

___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto, line 31 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7#-MXdGxRTA7Ps8hMsWGd6:-MXh9vA6-rAPLdabbXj9:b8aikmu) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/962edf9669bcb1a58dcd9e6a4e31eac06c700372/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto#L31)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

The Exchange PK is external_recurring_exchange_id plus Date. Does that seem reasonable?
</blockquote></details>

I'd prefer PKs made up of explicit IDs with the uniqueness property documented, but as long as you document what fields make up the PK I won't block.
___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto, line 37 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7#-MXdHHMA0zwkcDq_m0Wn:-MXhAwLLBnBB1ZtDGyNP:b-gxe3n3) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/962edf9669bcb1a58dcd9e6a4e31eac06c700372/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto#L37)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

No, it's arbitrary. Is starting at 1 and using attempt_number more canonical?
</blockquote></details>

Not sure about canonical, but if the field were just called `attempt`, I would assume it was a 1-based counter. Another data point is that we use `attempt` or `attempt_number` for all such existing fields in CMMS internal proto messages.

**Review by efoxepstein — COMMENTED** [review]
> Dismissed @SanjayVas from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @AlekseiS, @oliver-amzn, and @zasadnyy)

___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto, line 31 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7#-MXdGxRTA7Ps8hMsWGd6:-MXhDyyK5F3q4t9oYi2P:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/962edf9669bcb1a58dcd9e6a4e31eac06c700372/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto#L31)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

I'd prefer PKs made up of explicit IDs with the uniqueness property documented, but as long as you document what fields make up the PK I won't block.
</blockquote></details>

Done.
___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto, line 37 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7#-MXdHHMA0zwkcDq_m0Wn:-MXhE3KF4CqqonKbD_GM:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/962edf9669bcb1a58dcd9e6a4e31eac06c700372/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto#L37)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Not sure about canonical, but if the field were just called `attempt`, I would assume it was a 1-based counter. Another data point is that we use `attempt` or `attempt_number` for all such existing fields in CMMS internal proto messages.
</blockquote></details>

Done.

**Review by SanjayVas — APPROVED** [review]
> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @AlekseiS, @oliver-amzn, and @zasadnyy)

**Review by kungfucraig — APPROVED** [review]

File: src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+package wfa.measurement.internal.kingdom;
+
+import "google/protobuf/timestamp.proto";
+
+option java_package = "org.wfanet.measurement.internal.kingdom";
+option java_multiple_files = true;
+
+// Stores metadata about an ExchangeStepAttempt.
+message ExchangeStepAttemptDetails {
+  enum State {
```
Comment (stevenwarejones) [inline src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto:26]: how is this state different than the `ExchangeStepAttempt` state? https://github.com/world-federation-of-advertisers/cross-media-measurement/pull/10/files#diff-70da6a8230c7350d91d41dd49c001d173f912fc7553fc81b3ddfcab8cb711879R47

**Review by efoxepstein — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7)* status: all files reviewed, 1 unresolved discussion (waiting on @AlekseiS, @oliver-amzn, @stevenwarejones, and @zasadnyy)

___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto, line 26 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7#-MXdGU72doxHXBexv3XN-r3-26:-MYCLUGC15zujKPYCfR8:b-3wl283) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/389eb1ee09d08545c460fd4be43fe496fd7dc6f0/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto#L26)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

how is this state different than the `ExchangeStepAttempt` state? https://github.com/world-federation-of-advertisers/cross-media-measurement/pull/10/files#diff-70da6a8230c7350d91d41dd49c001d173f912fc7553fc81b3ddfcab8cb711879R47
</blockquote></details>

You're looking into the future... that CL fixes up some issues in this one (and builds on it). Somewhere in the UI it should say that that pull request is to be merged into the branch hosting this one. GitHub is pretty dreadful at dependent PRs compared to internal tooling so maybe this isn't the right way to do it?

File: src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto (hunk 11 above)
Comment (stevenwarejones) [reply src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto:26]: > > _[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7)_ status: all files reviewed, 1 unresolved discussion (waiting on @AlekseiS, @oliver-amzn, @stevenwarejones, and @zasadnyy)
> 
> _[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto, line 26 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7#-MXdGU72doxHXBexv3XN-r3-26:-MYCLUGC15zujKPYCfR8:b-3wl283) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/389eb1ee09d08545c460fd4be43fe496fd7dc6f0/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto#L26)):_
> 
> Previously, stevenwarejones (Steven Ware Jones) wrote…
> You're looking into the future... that CL fixes up some issues in this one (and builds on it). Somewhere in the UI it should say that that pull request is to be merged into the branch hosting this one. GitHub is pretty dreadful at dependent PRs compared to internal tooling so maybe this isn't the right way to do it?



> > _[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7)_ status: all files reviewed, 1 unresolved discussion (waiting on @AlekseiS, @oliver-amzn, @stevenwarejones, and @zasadnyy)
> 
> _[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto, line 26 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/7#-MXdGU72doxHXBexv3XN-r3-26:-MYCLUGC15zujKPYCfR8:b-3wl283) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/389eb1ee09d08545c460fd4be43fe496fd7dc6f0/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt_details.proto#L26)):_
> 
> Previously, stevenwarejones (Steven Ware Jones) wrote…
> You're looking into the future... that CL fixes up some issues in this one (and builds on it). Somewhere in the UI it should say that that pull request is to be merged into the branch hosting this one. GitHub is pretty dreadful at dependent PRs compared to internal tooling so maybe this isn't the right way to do it?

I'll put the comment on that pr

**Review by stevenwarejones — COMMENTED** [review]
lets get https://github.com/world-federation-of-advertisers/cross-media-measurement/pull/10 merged into this branch before we merge to master.

**Comment by SanjayVas** [comment]
> lets get #10 merged into this branch before we merge to master.

We only ever merge into the current working branch (in this case, `main`). That is to say there's no intent to ever merge one PR branch into another. The only reason that you'll see one PR with another PR branch listed as that it's intended to merge into `main` _after_ this one.

Basically, we want to keep small, focused PRs but also be able to send each PR out for review regardless if the PR it depends on has been merged.

**Review by stevenwarejones — APPROVED** [review]

**Pull request merged**